docs/designs/ps_design.md

```diff
@@ -0,0 +1,424 @@
+
+```python
+service PServer{
+    rpc pull_embedding_vector(Tensor) returns (Tensor);
+}
+```
+
+### Push Gradients
+After backward-pass, workers push gradients to PS.
+
+```python
+message PushGradientRequest{
+    int32 model_version = 1;
+    repeated Tensor gradients = 2;
+}
+message PushGradientResponse{
+    bool accepted = 1;
+    int32 model_version = 2;
+}
+service PServer{
+    rpc push_gradient(PushGradientRequest) returns (PushGradientResponse);
+}
+```
+
+### PS reports submodel version to the master
+The master needs to know the model version to decide when to save checkpoint and when to evaluate model. PS regularly reports the version of the submodel it possessed to the master.
+
+Please note different pserver has different submodel version. The master choose the maximum of these submodel versions as the current model version.
+
+```python
+message ReportSubmodelVersionRequest{
+    int32 model_version = 1;
+}
+service PServer{
+    rpc report_submodel_version(ReportSubmodelVersionRequest) returns (google.protobuf.Empty);
+}
+```
+
+### The master tells PS to save checkpoint
+When the master decides to save checkpoint, the master tells all the pservers to save checkpoint. Every pserver saves the submodel it possessed into a separate file.
+
+```python
+service PServer{
+    rpc save_checkpoint(google.protobuf.Empty) returns (google.protobuf.Empty);
+}
+```
+
+## Embedding Replicas in PS
+An ElasticDL job has *N* PS nodes. Embedding vectors are partitioned into these PS nodes. The user provides *M*, the number of replicas for embedding vectors. *M* must be smaller than *N* as each PS node uses other PS nodes to store its embedding replicas.
+
+Assume *E(i)* is the embedding vectors in PS node *PS(i)*, it has *M* replicas which are stored in PS nodes from *P((i + 1) % N)* to *P((i + M) % N)*. Also, *PS(i)* has *M* other PS node replicas from *E((i - M) % N)* to *E((i - 1) % N)*. 
+
+*PS(i)* maintains *M* updated embedding vector key sets *UKS_i(j) for j from 0 to M - 1*. When *PS(i)* sparsely updates its embedding vectors *E(i)*, it also add the updated embedding vector keys into these *M* sets. 
+
+
+*PS(i)* also periodically synchronize the replicas stored in it from PS nodes *PS((i - M) % N)* to *PS((i - 1) % N)*. The synchronization frequency can be several seconds.
+
+Each PS will provide a gRPC service for the replica synchronization.
+
+```
+message SynchronizeEmbeddingRequest {
+    int32 replica_index = 1;
+}
+
+message SynchronizeEmbeddingResponse {
+    map<string, Tensor> embedding_vectors = 1;
```

Comment:
The is a `name` field in `Tensor`, so we may use

```proto
repeated Tensors embedding_vectors = 1;
```

Reply:
OK. I will modify this proto after discussion.